plutus-ledger-api/src/PlutusLedgerApi/V1/Interval.hs

```diff
@@ -79,14 +80,30 @@ instance Pretty a => Pretty (Extended a) where
     pretty PosInf     = pretty "+∞"
     pretty (Finite a) = pretty a
 
+-- See Note [Enumerable Intervals]
 -- | Whether a bound is inclusive or not.
 type Closure = Bool
 
 -- | The upper bound of an interval.
 data UpperBound a = UpperBound (Extended a) Closure
-    deriving stock (Haskell.Eq, Haskell.Ord, Haskell.Show, Generic)
+    deriving stock (Haskell.Show, Generic)
     deriving anyclass (NFData)
 
+-- | For an enumerable type, turn an upper bound into a single inclusive
+-- bounding value.
+--
+-- Since the type is enumerable, non-inclusive bounds are equivalent
+-- to inclusive bounds on the predecessor.
+--
+-- See Note [Enumerable Intervals]
+strictUpperValue :: Enum a => UpperBound a -> Extended a
+-- already inclusive
+strictUpperValue (UpperBound v True)           = v
+-- take pred
+strictUpperValue (UpperBound (Finite x) False) = Finite $ pred x
```

Comment:
`pred` is partial - undefined on `minBound`. So it should return Maybe, but that's probably more costly, so I'm fine with documenting the partiality.

Reply:
We don't have fixed-size integers, so that's less of a problem. In fact I think we don't have any partial `succ` or `pred` definitions. Still a bit suspect I agree.

Reply:
> We don't have fixed-size integers, so that's less of a problem. In fact I think we don't have any partial `succ` or `pred` definitions. Still a bit suspect I agree.

We have partial `Enum Bool`:  `succ True` and `pred False` ?

Reply:
You are absolutely right, and so is the instance for `()`.

Reply:
Isn't this equivalent and simpler? 

```
strictUpperValue (UpperBound (Finite x) False) = Finite $ pred x
strictUpperValue v = v
```

Reply:
Perhaps a better function name: `discretizeUpperBound` or `normalizeUpperBound`?

Reply:
Also we would have a problem with `Enum ()` and `Enum Ordering` in PlutusTx

Reply:
> Isn't this equivalent and simpler?

It's equivalent but I think I prefer my version because the fall-through case is okay for two distinct reasons.